src/anti_hoisting.rs

```diff
@@ -0,0 +1,172 @@
+    pub trigger: Regex, // TODO only single letters are allowed
+    /// The output message to send to the user
+    /// Must contain the literal `${nickname}`
+    /// Example: `Box<${nickname}>`
+    pub output: String,
+    /// Roles that are able to bypass anti-hoisting
+    pub ignore_roles: Vec<Id<RoleMarker>>
+}
+
+/// Deserialize Configuration
+/// This is needed to provide config error handling at the time of deserialization
+impl<'de> Deserialize<'de> for Configuration {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de> {
+
+        #[derive(serde::Deserialize)]
+        struct RawConfiguration {
+            #[serde(default)]
+            enabled: bool,
+            #[serde(default = "default_max_nickname_length")]
+            max_nickname_length: u32,
+            #[serde(default = "default_min_nickname_length")]
+            min_nickname_length: u32,
+            #[serde(deserialize_with = "deserialize_regex")]
+            trigger: Regex,
+            output: String,
+            #[serde(default)]
+            ignore_roles: Vec<Id<RoleMarker>>
+        }
+
+        let RawConfiguration {
+            enabled,
+            max_nickname_length,
+            min_nickname_length,
+            trigger,
+            output,
+            ignore_roles
+        } = RawConfiguration::deserialize(deserializer)?;
+
+        let output = output.trim();
+        
+        // invalid if output does not contain ${nickname}
+        if !output.contains(AntiHoisting::NICKNAME_PLACEHOLDER) {
+            return Err(Error::custom("`output` must contain `${nickname}`"));
+        };
+
+        //TODO validation
+
+        Ok(Configuration {
+            enabled,
+            max_nickname_length,
+            min_nickname_length,
+            trigger,
+            output: output.to_owned(),
+            ignore_roles,
+        })
+    }
+}
+
+pub struct AntiHoisting {}
+
+impl AntiHoisting {
+    pub const INTENTS: Intents = Intents::GUILD_MEMBERS;
+    pub const NICKNAME_PLACEHOLDER: &str = "${nickname}";
+
+    /// run the main anti-hoisting logic
+    pub async fn serve(config: Configuration, mut events: Receiver<Arc<Event>>, http: Arc<Client>) {
+        loop {
+            let event = events.recv().await;
+
+            // match events where a member's name is hoisted
+            let hoisted_member = match event.as_deref() {
+                Err(RecvError::Closed) => return,
+                Err(_) => continue,
+                Ok(Event::MemberUpdate(m)) if m.nick.as_ref().map_or(false, |nick| AntiHoisting::is_hoisted(&config.trigger, &nick))  => m,
+                Ok(_) => continue,
+            };
+
+            // skip member with a role that bypasses anti-hoisting
+            if hoisted_member.roles.iter().any(|role| config.ignore_roles.contains(role)) {
+                debug!("Member {:#?} has a role that bypasses anti-hoisting, skipping", hoisted_member.user.name);
```

Comment:
```suggestion
                debug!("Member has a role that bypasses anti-hoisting, skipping", name = %hoisted_member.user.name);
```
see https://docs.rs/tracing/latest/tracing/index.html#using-the-macros

Reply:
Does the field not go in front of the message?